README.md

```diff
@@ -0,0 +1,237 @@
+            .build();
+
+    @Test
+    public void doSomethingWithMysqlTest() {
+        final String username = mysql.getUsername();
+    }
+```
+
+
+
+We can again accomplish a similar API to Testcontainers 1.x. The example above illustrates a getter for JDBC username on a MySql-specialized class.
+
+
+
+### Implementation
+
+Note that:
+
+* To avoid the mental overhead of generics that current exists in Testcontainers 1.x, here the `MySqlContainerTestRule` class is *not* a subclass of `ContainerTestRule`. Instead, it inherits a default JUnit `apply` method from the `TestRuleImplementation` interface.
+* As with `ContainerTestRule`, the `MySqlContainerTestRule` class _delegates_ to a `Container` instance using the Lombok `@Delegate` annotation.
+* `MySqlContainerTestRule` does some further composition to customize the `Container` instance without resorting to subclassing - it binds a `Plugin` . In this example, the MySql-specific plugin class could do things such as setting environment variables and selecting a readiness check strategy. The next section descibes what a `Plugin` is!
+
+
+
+---
+
+## Plugins
+
+Plugins are intended to allow extra behaviour to be added to `Container` via composition rather than inheritance. This should make it easier to share small, reusable, elements without tying them in to a particular branch of an inheritance hierarchy.
+
+### End result / usage
+
+This is draft and highly subject to change, but the Plugin API could look like the following.
+
+For the developer, adding a plugin to a container would be as simple as calling the `withPlugin` method on the builder, e.g. for a fictitious plugin that adds FakeTime to containers:
+
+```java
+final Container containerWithPlugin = Container.builder()
+  .withImage("alpine:3.5")
+  .withPlugin(new FakeTimePlugin("2010-01-01 00:00:00"))
+  .build();
+```
+
+A plugin would implement one or more of the methods on the `Plugin` interface (this is very incomplete, lacking parameters for example):
+
+```java
+public interface Plugin {
+
+    // draft list of hook points
+
+    default void beforeConfigure() {}
+    default void afterConfigure() {}
+
+    default void beforeStart() {}
+    default void afterStart() {}
+
+    default void waitForStarted() {}
+    default void waitForReady() {}
+
+    default void afterExit() {}
+}
+```
+
+Plugins would be called in order at the appropriate points in a container's lifecycle.
+
+Additionally it may be useful to allow the developer to write code that interacts with a plugin while containers are running. This could be debated, but perhaps the simplest model that could work is to have the test class hold a field reference to the plugin, e.g.:
+
+```java
+final FaketimePlugin fakeTime = new FakeTimePlugin("2010-01-01 00:00:00")
+
+final Container containerWithPlugin = Container.builder()
+  .withImage("alpine:3.5")
+  .withPlugin(fakeTime)
+  .build();
+
+// later, while container is running
+fakeTime.setTime("2011-01-01 00:00:00");
```

Comment:
would make it hard to reuse. How about `containerWithPlugin.using(fakeTime).setTime(...)`?

Reply:
What's the difference, since we still need the reference to the field?

Reply:
```java
containerFoo.using(fakeTime).setTime(...)
containerBar.using(fakeTime).setTime(...)
```

Reply:
Would be nice, might be problematic for stateful Plugins (which might exist).

Or maybe we could have something like:

```
final Container containerWithPlugin = Container.builder()
  .withImage("alpine:3.5")
  .withPlugin(FakeTimePlugin.class);

containerWithPlugin.using(FakeTimePlugin.class).setTime(...);

Reply:
Stateful plugins is a pain. Even if plugin wants to store the state, it's better if it will use something like https://docs.oracle.com/javase/7/docs/api/java/util/WeakHashMap.html